### PR TITLE
Fix/1217 create mocking lesson

### PR DESCRIPTION
**Pull request by pa7ri**
**Closes** #1217 

- Extract relevant information about mocking from [integration-testing](https://github.com/Techtonica/curriculum/blob/8acd5068d6bcbfb4f4c71ef598340bb324793764/testing-and-tdd/integration-testing.md#2-mocking-and-abstraction) lesson.
- Create a mocking lesson with the proper [template](https://github.com/Techtonica/curriculum/blob/master/_templates/topic-outline.md).
- Remove duplicated info from [integration-testing](https://github.com/Techtonica/curriculum/blob/8acd5068d6bcbfb4f4c71ef598340bb324793764/testing-and-tdd/integration-testing.md#2-mocking-and-abstraction) lesson.
- Add supplemental material for mocking & abstraction.

### CODE REVIEW

File: testing-and-tdd/integration-testing.md [hunk 1]
```diff
@@ -210,210 +205,11 @@ use while writing unit tests for your project and then dig into concrete
 examples in a small node.js + express project and corresponding tests.
 
 How you consume this will depend your learning style. My suggestion is to skim
-over _Mocking_ and _Abstraction_ sections then work through the examples. It's
+over [_Mocking and Abstraction_][tt-mocking] section then work through the examples. It's
 pretty normal for writing code with modularity and testing in mind so don't
 fret if it takes more than a couple of passes for things to fall into place.
```
Comment (Kundan28) [inline testing-and-tdd/integration-testing.md]: Why these are three separate lines ? It should be one line instead.

File: testing-and-tdd/mocking-and-abstraction.md [hunk 2]
```diff
@@ -0,0 +1,367 @@
+# Mocking and Abstraction
+
+### Projected Time
+
+About 1 hour and 10 minutes
```
Comment (Kundan28) [inline testing-and-tdd/mocking-and-abstraction.md]:
```suggestion
About 1 hour and 20 minutes
```

File: testing-and-tdd/mocking-and-abstraction.md [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+- Lesson: 30 min
+- Guided Practice: 15 min
+- Independent Practice: 25 min
+- Check for Understanding: 10 min
+
+
+[simplemock-home]: https://www.npmjs.com/package/simple-mock
+[nock-home]: https://github.com/nock/nock
+[nock-intro]: https://scotch.io/tutorials/nodejs-tests-mocking-http-requests
+[backend-ii]: https://glitch.com/edit/#!/zee-techtonica-backend-testing-ii?path=index.js:37:0
```
Comment (Kundan28) [inline testing-and-tdd/mocking-and-abstraction.md:16]:
```suggestion
[simplemock-home](https://www.npmjs.com/package/simple-mock)
[nock-home](https://github.com/nock/nock)
[nock-intro](https://scotch.io/tutorials/nodejs-tests-mocking-http-requests)
[backend-ii](https://glitch.com/edit/#!/zee-techtonica-backend-testing-ii?path=index.js:37:0)
```

File: testing-and-tdd/mocking-and-abstraction.md [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+
+Here are links to lessons that should be completed before this lesson:
+
+- [Introduction to Testing][tt-testing-intro]
+- [Jasmine Testing][tt-testing-frameworks]
+- [Integration Testing][tt-integration-testing]
+
+[tt-testing-intro]: /testing-and-tdd/testing-and-tdd.md
+[tt-testing-frameworks]: jasmine-testing.md
+[tt-integration-testing]: ntegration-testing.md
```
Comment (Kundan28) [inline testing-and-tdd/mocking-and-abstraction.md:28]: Fix these too as I said above.

File: testing-and-tdd/mocking-and-abstraction.md [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+### Motivation
+
+Continuing with our testing sections, we will explore here that the core of our tests will be built on the concept of providing mocked responses to external service calls. This allows us to take control over much of the complexity of interacting with other services. It additionally helps address the potential time and money costs
+that making actual calls to the service would introduce into our tests.
```
Comment (Kundan28) [inline testing-and-tdd/mocking-and-abstraction.md]: Make these in one line too,

File: testing-and-tdd/mocking-and-abstraction.md [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+- Use mocks to mimic testing GET requests to an external API.
+- Use Abstraction to simplify testing.
+
+### Specific Things To Learn
+
+Testing external services
+- Mocking & abstractions
+- [Simple Mock][simplemock-home]
+- [Nock][nock-home] ([intro tutorial][nock-intro])
```
Comment (Kundan28) [inline testing-and-tdd/mocking-and-abstraction.md:47]: Fix these too.

File: testing-and-tdd/mocking-and-abstraction.md [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+
+- [Example video (10 min)](https://google.com) - Write a very short description of this resource.
+- [Other example article(20 min read)](https://google.com) - Answer the 3 questions at the end.
+  (Be sure to explain how to use these materials in directions in one of the sections below (ie: lesson, guided practice). If you are not assigning them somehow, they should be placed under Supplemental Materials at the bottom.)
+
+### Lesson
+
+#### Mocking
+
+The concept of **mocking** was covered in [Intro to Testing][tt-testing-intro].
```
Comment (Kundan28) [inline testing-and-tdd/mocking-and-abstraction.md:59]: This link too.

File: testing-and-tdd/mocking-and-abstraction.md [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+
+> As a brief refresher: it is a technique of providing an implementation of an
+> interface which allows you to specify exactly what the return value should be
+> when a specific call is made. Additionally, it enables you to verify that the
+> interface was called with the expected values.
+
+In order to mock backend calls we'll be using a library called [`nock`][nock-home]. Nock works by intercepting HTTP requests that your code makes checking against what you've instructed it to expect. If it finds a match it will return the response you've configured, if not it will result in a test failure.
```
Comment (Kundan28) [inline testing-and-tdd/mocking-and-abstraction.md:66]: This link too.

File: testing-and-tdd/mocking-and-abstraction.md [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+- Call `getUser('not-octocat')`?
+- Change `mockObject.id` to be `42`?
+- Change `mockObject.name` to `Techtonica`?
+
+#### Abstraction
+
+Think back to [Eloquent JavaScript Ch 5][ejs-5] when you learned about
+_Abstraction_ and _Higher-order Functions_. Recall that these techniques are used to wrap reptitive or complex behavior and then provide a more easily understandable way to access that behavior. When thinking about how to unit test your project we'll be making heavy use of these concepts. We do so to create functions that are as simple as possible so that the tests we write don't get too complex.
+
+[ejs-5]: https://eloquentjavascript.net/05_higher_order.html
```
Comment (Kundan28) [inline testing-and-tdd/mocking-and-abstraction.md:130]: Fix this issue everywhere.

File: testing-and-tdd/mocking-and-abstraction.md [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+
+    // verify that functionToTest only passed one parameter
+    expect(callArgs.length).to.equal(1);
+    // ...and that the parameter's value was 1
+    expect(callArgs.length[0]).to.equal(1);
+  });
+});
+```
+
+This is enough for you to get a solid collection of tests going for the code that calls your database but `simple-mock` is much more featureful and it's worth looking into the different testing / validation modes it supports later.
```
Comment (Kundan28) [inline testing-and-tdd/mocking-and-abstraction.md:350]:
```suggestion
This is enough for you to get a solid collection of tests going for the code that calls your database but `simple-mock` is much more featureful and it's worth looking into the different testing/validation modes it supports later.
```

File: testing-and-tdd/mocking-and-abstraction.md (hunk 3 above)
Comment (pa7ri) [reply testing-and-tdd/mocking-and-abstraction.md:16]: mmm I agree with the rest of the changes except for this one, im not calling a link here, im defining it to use it later on

Comment (Kundan28) [reply testing-and-tdd/mocking-and-abstraction.md:16]: Oh, okay. Then you could leave it like that @pa7ri

File: testing-and-tdd/mocking-and-abstraction.md [hunk 11]
```diff
@@ -124,18 +122,15 @@ Following example above, try to represent the following scenarios and think abou
...
 
 
 **An Example:**
-Let's look at some places where abstraction can help us make our code easier
-to understand and maintain.
+Let's look at some places where abstraction can help us make our code easier to understand and maintain.
 
-In the following code snippet we're working in a basic express app that can
-list and add items to a TODO list:
+In the following code snippet we're working in a basic express app that can list and add items to a TODO list:
```
Comment (Kundan28) [inline testing-and-tdd/mocking-and-abstraction.md:133]:
```suggestion
In the following code snippet we're working in a basic express app that can list and add items to a To Do list:
```

**Comment by Kundan28** [comment]
Updated in the scoresheet too.

**Pull request merged**